mingpt/fake_lightning.py

```diff
@@ -0,0 +1,163 @@
+"""
+A manual, minimal and non-full-featured implementation of boilerplate training loop.
+Intentionally made to have the same API as PyTorch Lightning, giving two benefits:
+1) Everyone can inspect/hack this simple implementation for educational purposes
+2) Everyone can run the full Lightning implementation when they just want to go FAST
+"""
+
+import os
+import math
+import logging
+
+from tqdm import tqdm
+import torch
+import torch.nn as nn
+
+logger = logging.getLogger(__name__)
+
+# -----------------------------------------------------------------------------
+
+class Result:
+        """ very thin wrapper around a result of a train/val/test step of the model """
+        def __init__(self, minimize=None, checkpoint_on=None):
+            self.minimize = minimize
+            self.checkpoint_on = checkpoint_on
+
+        def log(self, key, val):
+            setattr(self, key, val)
+
+class TrainResult(Result):
+    pass
+
+class EvalResult(Result):
+    pass
+
+class LightningModule(nn.Module):
+
+    def load_from_checkpoint(self, checkpoint_path):
+        logger.info("loading the best model checkpoint from %s", checkpoint_path)
+        state_dict = torch.load(checkpoint_path)
+        self.load_state_dict(state_dict)
```

Comment:
In lightning the checkpoint is not just the weights.

The checkpoint also has:
- all the training state (epoch, state, etc)
- hyperparams used in training
- the lightning version
- states of 16-bit scaling, etc
- states of all the learning rate schedulers, etc...

Load from checkpoint restores all that. 

If you want just the state dict then you can do:
```
ckpt = torch.load(path)
model.load_state_dict(ckpt['state_dict'])
```